test/common/config/metadata_test.cc

```diff
@@ -51,76 +51,213 @@ TEST(MetadataTest, MetadataValuePath) {
+        registered_factory_baz_(baz_factory_) {}
 
   struct Foo : public TypedMetadata::Object {
     Foo(std::string name) : name_(name) {}
     std::string name_;
   };
 
-  struct Bar : public TypedMetadata::Object {};
+  struct Qux : public TypedMetadata::Object {};
 
-  class FooFactory : public TypedMetadataFactory {
+  class FoobarFactory : public TypedMetadataFactory {
   public:
-    std::string name() const override { return "foo"; }
     // Throws EnvoyException (conversion failure) if d is empty.
     std::unique_ptr<const TypedMetadata::Object>
     parse(const ProtobufWkt::Struct& d) const override {
       if (d.fields().find("name") != d.fields().end()) {
         return std::make_unique<Foo>(d.fields().at("name").string_value());
       }
-      throw EnvoyException("Cannot create a Foo when metadata is empty.");
+      throw EnvoyException("Cannot create a Foo when Struct metadata is empty.");
+    }
+
+    std::unique_ptr<const TypedMetadata::Object> parse(const ProtobufWkt::Any& d) const override {
+      if (!(d.type_url().empty())) {
+        return std::make_unique<Foo>(d.value());
+      }
+      throw EnvoyException("Cannot create a Foo when Any metadata is empty.");
+    }
+  };
+
+  // Three testing factories with different names
+  class FooFactory : public FoobarFactory {
+  public:
+    std::string name() const override { return "foo"; }
+  };
+
+  class BarFactory : public FoobarFactory {
+  public:
+    std::string name() const override { return "bar"; }
+  };
+
+  class BazFactory : public FoobarFactory {
+  public:
+    std::string name() const override { return "baz"; }
+    // Override Any parse() to just return nullptr.
+    std::unique_ptr<const TypedMetadata::Object> parse(const ProtobufWkt::Any&) const override {
+      return nullptr;
     }
   };
 
 protected:
   FooFactory foo_factory_;
-  Registry::InjectFactory<TypedMetadataFactory> registered_factory_;
+  BarFactory bar_factory_;
+  BazFactory baz_factory_;
+  Registry::InjectFactory<TypedMetadataFactory> registered_factory_foo_;
+  Registry::InjectFactory<TypedMetadataFactory> registered_factory_bar_;
+  Registry::InjectFactory<TypedMetadataFactory> registered_factory_baz_;
 };
 
-TEST_F(TypedMetadataTest, OkTest) {
+TEST_F(TypedMetadataTest, OkTestStruct) {
   envoy::config::core::v3::Metadata metadata;
   (*metadata.mutable_filter_metadata())[foo_factory_.name()] =
-      MessageUtil::keyValueStruct("name", "foo");
+      MessageUtil::keyValueStruct("name", "garply");
   TypedMetadataImpl<TypedMetadataFactory> typed(metadata);
   EXPECT_NE(nullptr, typed.get<Foo>(foo_factory_.name()));
-  EXPECT_EQ("foo", typed.get<Foo>(foo_factory_.name())->name_);
+  EXPECT_EQ("garply", typed.get<Foo>(foo_factory_.name())->name_);
   // A duck is a duck.
-  EXPECT_EQ(nullptr, typed.get<Bar>(foo_factory_.name()));
+  EXPECT_EQ(nullptr, typed.get<Qux>(foo_factory_.name()));
+}
+
+TEST_F(TypedMetadataTest, OkTestAny) {
```

Comment:
Can you please add a comment before each test describing what the test verifies.

Reply:
Yes, good idea! Done.